WalletWasabi/Wallets/FileSystemBlocksRepository.cs

```diff
@@ -0,0 +1,133 @@
+using NBitcoin;
+using NBitcoin.DataEncoders;
+using Nito.AsyncEx;
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Wallets
+{
+	/// <summary>
+	/// FileSystemBlocksRepository is a blocks repository that keep the blocks in the file system.
+	/// </summary>
+	public class FileSystemBlocksRepository : IRepository<uint256, Block>
+	{
+		public FileSystemBlocksRepository(string blocksFolderPath, Network network)
+		{
+			BlocksFolderPath = blocksFolderPath;
+			Network = network;
+			CreateFolders();
+		}
+
+		public string BlocksFolderPath { get; }
+		public Network Network { get; }
+		private AsyncLock BlockFolderLock { get; } = new AsyncLock();
+
+		/// <summary>
+		/// Gets a bitcoin block from the file system.
+		/// </summary>
+		/// <param name="hash">The block's hash that identifies the requested block.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>The requested bitcoin block.</returns>
+		public async Task<Block> GetAsync(uint256 hash, CancellationToken cancellationToken)
+		{
+			// Try get the block
+			Block block = null;
+			using (await BlockFolderLock.LockAsync())
+			{
+				var encoder = new HexEncoder();
+				var filePath = Path.Combine(BlocksFolderPath, hash.ToString());
+				if (File.Exists(filePath))
+				{
+					try
+					{
+						var blockBytes = await File.ReadAllBytesAsync(filePath, cancellationToken);
+						block = Block.Load(blockBytes, Network);
+					}
+					catch
+					{
+						// In case the block file is corrupted and we get an EndOfStreamException exception
```

Comment:
```suggestion
						// In case the block file is corrupted and we get an EndOfStreamException
```

Reply:
this is okay as it is and I don't want to touch anything because it is already hard to review.